src/psyclone/domain/lfric/algorithm/alg_gen.py

```diff
@@ -0,0 +1,416 @@
+    kernel_routine = table.new_symbol(kernel_name,
+                                      symbol_type=DataTypeSymbol,
+                                      datatype=DeferredType(),
+                                      interface=ImportInterface(kernel_mod))
+
+    ktype = KernelTypeFactory(api="dynamo0.3").create(parse_tree,
+                                                      name=kernel_name)
+    # Construct a DynKern using the metadata. This is used when constructing
+    # the kernel argument list.
+    kern = DynKern()
+    kern.load_meta(ktype)
+
+    # Declare and initialise the data structures required by the kernel
+    # arguments. Appropriate symbols are added to the symbol table associated
+    # with the routine we are constructing.
+    kern_args = construct_kernel_args(sub, kern)
+
+    # Initialise argument values to unity. Since we are using this somewhat
+    # arbitrary value, we use an *integer* literal for this, irrespective of
+    # the actual type of the scalar argument. The compiler/run-time will take
+    # care of appropriate type casting.
+    psyir.add_lfric_precision_symbol(table, "i_def")
+    for sym in kern_args.scalars:
+        sub.addchild(Assignment.create(
+            Reference(sym),
+            Literal("1", psyir.LfricIntegerScalarDataType())))
+
+    # We use the setval_c builtin to initialise all fields to unity.
+    # Currently we have to create a symbol for this builtin in order to
+    # create the appropriate functor.
+    # TODO #1645 - we do *not* put this symbol in the table as otherwise
+    # the backend will currently fail because it is undefined.
+    setval_c = DataTypeSymbol("setval_c", DeferredType())
+
+    # As with the scalar initialisation, we don't worry about precision
+    # here since we are just setting the field values to unity. If the
+    # field itself is of a precision other than r_def (or is perhaps
+    # integer rather than real) we rely on type casting by the
+    # compiler/run-time.
+    psyir.add_lfric_precision_symbol(table, "r_def")
+    kernel_list = []
+    for sym, _ in kern_args.fields:
+        kernel_list.append(
+            LFRicBuiltinFunctor.create(
+                setval_c, [Reference(sym),
+                           Literal("1.0", psyir.LfricRealScalarDataType())]))
+
+    # Finally, add the kernel itself to the list for the invoke().
+    arg_nodes = []
+    for arg in kern_args.arglist:
+        arg_nodes.append(Reference(table.lookup(arg)))
+    kern = LFRicKernelFunctor.create(kernel_routine, arg_nodes)
+    kernel_list.append(kern)
+
+    # Create the 'call invoke(...)' for the list of kernels.
+    invoke_sym = table.new_symbol("invoke", symbol_type=RoutineSymbol)
+    sub.addchild(LFRicAlgorithmInvokeCall.create(invoke_sym, kernel_list, 0))
+
+    return FortranWriter()(cont)
+
+
+# For automatic API documentation.
+__all__ = ["initialise_field", "initialise_quadrature",
```

Comment:
(There is also `src/psyclone/alg_gen.py` which does define `AlgGen`.)

Reply:
Sorry for being unclear: I typically prefer that a file only exports a single symbol, in this case a class with static functions. That has imho two advantages:

1. When a function is used, it will have the class name: `AlgGen.generate(...)`, as opposed to just `generate()` (which requires you to check the import statements to find where it is coming from - luckily our coding style prevents `import *` :) ).
2. It avoids potential name clashes (e.g. what if a function should need two  `generate` functions that are defined in two different files?)
 
Admittedly, 2 is rather weak (we import the file and then use `file.function()`, and the fact that we don't allow `import *` makes the argument of namespace pollution mostly moot :)  ). And if later you add a test-harness, it might allow for better design if you can either inherit from this class, or in this new class add an instance that allows access to the functions. 

But am happy to leave this, it's not a required coding style, and can very easily be refactored later if required.